src/hdskins/java/com/voxelmodpack/hdskins/skins/SkinServer.java

```diff
@@ -5,45 +5,46 @@
 import com.google.gson.GsonBuilder;
 import com.mojang.authlib.GameProfile;
 import com.mojang.authlib.exceptions.AuthenticationException;
-import com.mojang.authlib.minecraft.MinecraftProfileTexture;
 import com.mojang.authlib.minecraft.MinecraftSessionService;
 import com.mojang.authlib.yggdrasil.response.MinecraftTexturesPayload;
 import com.mojang.util.UUIDTypeAdapter;
 import com.mumfrey.liteloader.modconfig.Exposable;
-
 import net.minecraft.client.Minecraft;
 import net.minecraft.util.Session;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
 
-import java.net.URI;
-import java.util.Collections;
+import java.io.IOException;
 import java.util.List;
-import java.util.Map;
-import java.util.Optional;
 import java.util.UUID;
 import java.util.concurrent.CompletableFuture;
-
-import javax.annotation.Nullable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 
 public interface SkinServer extends Exposable {
 
-    static final Gson gson = new GsonBuilder()
+    Gson gson = new GsonBuilder()
             .registerTypeAdapter(UUID.class, new UUIDTypeAdapter())
             .create();
 
     List<SkinServer> defaultServers = Lists.newArrayList(new LegacySkinServer(
             "http://skins.voxelmodpack.com",
             "http://skinmanager.voxelmodpack.com"));
 
-    Optional<MinecraftTexturesPayload> loadProfileData(GameProfile profile);
+    CloseableHttpClient httpClient = HttpClients.createSystem();
 
-    default Map<MinecraftProfileTexture.Type, MinecraftProfileTexture> getPreviewTextures(GameProfile profile) {
-        return loadProfileData(profile).map(MinecraftTexturesPayload::getTextures).orElse(Collections.emptyMap());
-    }
+    ExecutorService skinDownloadExecutor = Executors.newFixedThreadPool(8);
+    ExecutorService skinUploadExecutor = Executors.newSingleThreadExecutor();
 
-    CompletableFuture<SkinUploadResponse> uploadSkin(Session session, @Nullable URI image, MinecraftProfileTexture.Type type, Map<String, String> metadata);
+    MinecraftTexturesPayload loadProfileData(GameProfile profile) throws IOException;
 
+    CompletableFuture<SkinUploadResponse> uploadSkin(Session session, SkinUpload upload);
+
+    default CompletableFuture<MinecraftTexturesPayload> getPreviewTextures(GameProfile profile) {
```

Comment:
Yes it does. It's needed because of the way legacy works. The gateway server doubles as a cache bypass. If we ever remove the legacy server support, then we can remove that.

Reply:
Hm. Okay then.

How does that work to bypass the cache?

Reply:
It's server-side. And I'm just speculating here based on the source which mumfrey gave me. I have no idea how the server or cloudflare is configured. it stays for now.